### PR TITLE
metrics(download): move header stage metrics to downloader

### PR DESCRIPTION
**Pull request by rkrasiuk**
The header stage error metrics were no longer updated, because the downloader doesn't return any errors any more.

This PR moves header metrics into the downloader.

There is one additional metric tracking the total number of downloaded headers (including invalid ones).

### CODE REVIEW

File: crates/net/downloaders/src/headers/metrics.rs [hunk 1]
```diff
@@ -0,0 +1,40 @@
+use metrics::Counter;
+use reth_interfaces::p2p::error::DownloadError;
+use reth_metrics_derive::Metrics;
+
+/// The header downloader metrics.
+#[derive(Metrics)]
+#[metrics(scope = "downloaders_headers")]
```
Comment (onbjerg) [inline crates/net/downloaders/src/headers/metrics.rs:7]: metrics will translate to `downloaders_headers` and recommends using `.` as a separator

```suggestion
#[metrics(scope = "downloaders.headers")]
```

File: crates/net/downloaders/src/headers/metrics.rs [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+pub struct HeaderDownloaderMetrics {
+    /// The number of headers that were successfully returned by the downloader.
+    total_returned: Counter,
```
Comment (onbjerg) [inline crates/net/downloaders/src/headers/metrics.rs:10]: Not sure about the naming and comment here. I understand what it is saying, but I think w/o my context it wouldn't make a lot of sense

Maybe `total_flushed`? The comment should at least specify that this is the number of headers sent to the stage.

File: crates/net/downloaders/src/headers/metrics.rs [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+    /// Number of headers that were successfully downloaded
+    total_downloaded: Counter,
+    /// Number of timeout errors while requesting headers
+    timeout_errors: Counter,
+    /// Number of validation errors while requesting headers
+    validation_errors: Counter,
+    /// Number of unexpected errors while requesting headers
+    unexpected_errors: Counter,
```
Comment (onbjerg) [inline crates/net/downloaders/src/headers/metrics.rs:18]: IIRC we got support for labels in `metrics-derive`, right? In that case I think it is nicer to just have an `errors` metric with different labels, e.g. `errors{type="timeout"}`/`errors{type="invalid_header"}` etc.

File: crates/net/downloaders/src/headers/metrics.rs [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+        self.total_returned.increment(number as u64)
+    }
+
+    /// Increment the number of total downloaded headers.
+    pub(crate) fn update_total_downloaded(&self, number: usize) {
+        self.total_downloaded.increment(number as u64)
+    }
+
+    /// Update header errors metrics
+    pub(crate) fn update_error_metrics(&self, error: &DownloadError) {
```
Comment (onbjerg) [inline crates/net/downloaders/src/headers/metrics.rs:33]: This one makes sense though
```suggestion
    pub(crate) fn increment_errors(&self, error: &DownloadError) {
```

**Review by onbjerg — CHANGES_REQUESTED** [review]
Few nits. We should also add metrics for the bodies downloader either here or in a separate PR. If we do it in a separate one and you don't want that PR we should track it in an issue

**Comment by rkrasiuk** [comment]
@onbjerg it will be a separate PR

File: crates/net/downloaders/src/headers/metrics.rs (hunk 1 above)
Comment (rkrasiuk) [reply crates/net/downloaders/src/headers/metrics.rs:7]: yeah, we need to revert that PR that prohibits usage of dots first

Comment (rkrasiuk) [reply crates/net/downloaders/src/headers/metrics.rs:7]: opened https://github.com/paradigmxyz/reth/pull/1018

**Review by mattsse — APPROVED** [review]
lgtm,

pedantic nit :)

File: crates/net/downloaders/src/headers/metrics.rs [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+use reth_metrics_derive::Metrics;
+
+/// The header downloader metrics.
+#[derive(Metrics)]
+#[metrics(scope = "downloaders_headers")]
+pub struct HeaderDownloaderMetrics {
+    /// The number of headers that were successfully sent to the poller (stage)
+    pub(crate) total_flushed: Counter,
+    /// Number of headers that were successfully downloaded
+    pub(crate) total_downloaded: Counter,
```
Comment (mattsse) [inline crates/net/downloaders/src/headers/metrics.rs:12]: pedantic style nit:

I don't like mixing accessors here, I guess we can make all of them `pub(crate)`?

File: crates/net/downloaders/src/headers/metrics.rs (hunk 3 above)
Comment (rkrasiuk) [reply crates/net/downloaders/src/headers/metrics.rs:18]: let's do it in a follow up. metrics label UX is a little bit scarce, we need to discuss & improve it

**Pull request merged**